tests/test_typechecker.py

```diff
@@ -4683,31 +4688,36 @@ class L(typing.List[T]):
         self.assertFalse(pytypes.is_subtype(L[float], C, bound_typevars_readonly=True, bound_typevars={}))
         self.assertTrue(pytypes.is_subtype(L[float], C, bound_typevars_readonly=False, bound_typevars={}))
 
+    # See: https://github.com/Stewori/pytypes/issues/22
     def test_forward_declaration(self):
-        Container = typing.Union[
+        Wrapper = typing.Union[
             typing.List['Data'],
         ]
 
         Data = typing.Union[
-            Container,
+            Wrapper,
             str, bytes, bool, float, int, dict,
         ]
 
         with self.assertRaises(pytypes.ForwardRefError):
-            pytypes.is_subtype(typing.List[float], Container)
+            pytypes.is_subtype(typing.List[float], Wrapper)
 
-        pytypes.resolve_fw_decl(Container)
+        pytypes.resolve_fw_decl(Wrapper)
 
-        pytypes.is_subtype(typing.List[float], Container)
+        self.assertTrue(pytypes.is_subtype(typing.List[float], Wrapper))
```

Comment:
This is where the test fails. Before the change, the result was not checked to be true, it succeeded whenever no exception was thrown. Actually ``List[float]`` is no subtype of ``Wrapper`` because of invariance of ``List`` 's type parameter. Either construct Wrapper differently or change expected outcome of this test. E.g.:

```
Wrapper = typing.Union[
             typing.Sequence['Data'],
         ]
```
Disclaimer: I din't check empirically. I hope my analysis is correct and this is the actual cause.

Reply:
How it is not? I can get there by simply rewriting:

`Wrapper` -> `Union[List[Data]]` -> `List[Data]` -> `List[Union[Wrapper, str, bytes, bool, float, int, dict]]` -> `List[float]`

So `List[float]` should be a subtype of `List[float]`, no?

```
>>> pytypes.is_subtype(typing.List[float], typing.List[float])
True
```

Reply:
``List[Union[Wrapper, str, bytes, bool, float, int, dict]]`` -> ``List[float]`` is not valid, because ``Union[Wrapper, str, bytes, bool, float, int, dict]`` is a less restrictive type than ``float`` and ``List`` is invariant. Imagine someone passes ``List[float]`` to a function that expects ``List[Union[Wrapper, str, bytes, bool, float, int, dict]]``. Then the method might write integers or strings into the list.
What is maybe a bad logic in ``deep_type`` is to conclude that something like ``[1.5, 2.3, 5.7]`` is ``List[float]``. Just because is doesn't contain strings does not mean it is not intended to ever do so. This is a symptom of PEP 484 not being intended for runtime type checking. But python does not provide a way to annotate that a list object that currently contains only floats is also allowed to contain strings. (Except PEP 526 -- Syntax for Variable Annotations, but supporting PEP 526 is not a trivial enhanchement.) Some simpler solution would be needed.... Ideas?

Reply:
Okay, since this test is about another topic, please change it to ``Sequence``. We can follow up on ``List`` invariance issues else where, e.g. in https://github.com/Stewori/pytypes/issues/21.

Reply:
Done.